MapboxNavigation/MapboxSpeechSynthesizer.swift

```diff
@@ -0,0 +1,252 @@
+    private var previousInstrcution: SpokenInstruction?
+    
+    // MARK: - Lifecycle
+    
+    init(_ accessToken: String? = nil) {
+        self.cache = DataCache()
+        self.speech = SpeechSynthesizer(accessToken: accessToken)
+    }
+    
+    deinit {
+        // stop talking and unduck
+    }
+    
+    // MARK: - Methods
+    
+    public func changedIncomingSpokenInstructions(_ instructions: [SpokenInstruction]) {
+        instructions
+            .prefix(stepsAheadToCache)
+            .forEach {
+                if !hasCachedSpokenInstructionForKey($0.ssmlText) {
+                    downloadAndCacheSpokenInstruction(instruction: $0)
+                }
+        }
+    }
+    
+    public func speak(_ instruction: SpokenInstruction, completion: SpeechSynthesizerCompletion?) {
+        if let data = cachedDataForKey(instruction.ssmlText) {
+            safeDuckAudio(instruction: instruction)
+            completion?(speakWithMapboxSynthesizer(instruction: instruction,
+                                              instructionData: data))
+        }
+        else {
+            self.completion = completion
+            fetchAndSpeak(instruction: instruction)
+        }
+    }
+    
+    public func stopSpeaking() {
+        audioPlayer?.stop()
+    }
+    
+    public func interruptSpeaking() {
+        audioPlayer?.stop()
+    }
+    
+    // MARK: - Private Methods
+    
+    /**
+     Fetches and plays an instruction.
+     */
+    private func fetchAndSpeak(instruction: SpokenInstruction){
+        audioTask?.cancel()
+        
+        let modifiedInstruction = delegate?.voiceController(self, willSpeak: instruction) ?? instruction
+        let ssmlText = modifiedInstruction.ssmlText
+        let options = SpeechOptions(ssml: ssmlText)
+        options.locale = locale
+                        
+        audioTask = speech.audioData(with: options) { [weak self] (data, error) in
+            defer { self?.completion = nil }
+            
+            guard let self = self else { return }
+            if let error = error,
+                case let .unknown(response: _, underlying: underlyingError, code: _, message: _) = error,
+                let urlError = underlyingError as? URLError, urlError.code == .cancelled {
+                self.completion?(error)
+                return
+            } else if let error = error {
+                self.delegate?.voiceController(self, spokenInstructionsDidFailWith: SpeechError.apiError(instruction: modifiedInstruction,
+                                                                                                    options: options,
+                                                                                                    underlying: error))
+                self.completion?(SpeechError.apiError(instruction: modifiedInstruction,
+                                                      options: options,
+                                                      underlying: error))
```

Comment:
Error propagation is still a TODO.

I feel the need to have explicit `completion` for `speak` method to allow it to by async, but at the same time we want to have a `delegate` to track various events. Also, some errors may be blocking for Speech Synthesizer to actually speak, others may be not.
What is the best way to handle errors in this situation? We can:
1. report any errors occur with via `delegate`, and `completion` will receive error only if it was blocking (or nil otherwise)
2. report any errors occur with via `delegate`, and `completion` will just have boolean `success`, and thus the last error reported was a blocker.
3. get rid of `completion` and make it a part of `delegate` protocol (with one of the 2 above points implemented)
4. something else?

Reply:
If we have anything bubble up to the delegate, it should be after the speech controller has finished processing all the speech synthesizers. So if there’s an error, try falling back to another speech synthesizer, and only notify the delegate if all the speech synthesizers fail.